File: an3/Semestrul1/LFTC/Lab/L2/BST.py
class TreeNode:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None

class BinarySearchTree:
    def __init__(self):
        self.root = None

    def insert(self, key):
        self.root = self._insert_recursive(self.root, key)

    def _insert_recursive(self, root, key):
        if root is None:
            return TreeNode(key)
        if key < root.key:
            root.left = self._insert_recursive(root.left, key)
        else:
            root.right = self._insert_recursive(root.right, key)
        return root
    
    def get_index(self, key):
        return self._get_index_recursive(self.root, key, 0)

    def _get_index_recursive(self, root, key, index):
        if root is None:
            return None
        if key == root.key:
            return index + self._count_nodes(root.left)
        if key < root.key:
            return self._get_index_recursive(root.left, key, index)
        return self._get_index_recursive(root.right, key, index + self._count_nodes(root.left) + 1)

    def _count_nodes(self, root):
        if root is None:
            return 0
        return 1 + self._count_nodes(root.left) + self._count_nodes(root.right)

    def inorder_traversal_print(self):
        return self._inorder_recursive_print_(self.root)

    def _inorder_recursive_print_(self, root):
        if root is not None:
            self._inorder_recursive_print_(root.left)
            print(root.key + " : " + str(self.get_index(root.key)))
            self._inorder_recursive_print_(root.right)

File: an3/Semestrul1/LFTC/Lab/L3/BST.c
#include <stdio.h>
#include <stdlib.h>

struct TreeNode {
    int key;
    struct TreeNode* left;
    struct TreeNode* right;
};

struct TreeNode* initializeTree() {
    return NULL; // an empty tree has a NULL root
}

struct TreeNode* createTreeNode(int key) {
    struct TreeNode* newNode = (struct TreeNode*)malloc(sizeof(struct TreeNode));
    newNode->key = key;
    newNode->left = NULL;
    newNode->right = NULL;
    return newNode;
}

struct TreeNode* insertRecursive(struct TreeNode* root, int key) {
    if (root == NULL) {
        return createTreeNode(key);
    }

    if (key < root->key) {
        root->left = insertRecursive(root->left, key);
    } else {
        root->right = insertRecursive(root->right, key);
    }

    return root;
}

int countNodes(struct TreeNode* root) {
    if (root == NULL) {
        return 0;
    }

    return 1 + countNodes(root->left) + countNodes(root->right);
}

int getIndexRecursive(struct TreeNode* root, int key, int index) {
    if (root == NULL) {
        return -1; // sau altă valoare de semnalare a lipsei cheii
    }

    if (key == root->key) {
        return index + countNodes(root->left);
    }

    if (key < root->key) {
        return getIndexRecursive(root->left, key, index);
    }

    return getIndexRecursive(root->right, key, index + countNodes(root->left) + 1);
}

void inorderRecursivePrint(struct TreeNode* root) {
    if (root != NULL) {
        inorderRecursivePrint(root->left);
        printf("%d : %d\n", root->key, getIndexRecursive(root, root->key, 0));
        inorderRecursivePrint(root->right);
    }
}

// int main() {
//     struct TreeNode* root = NULL;

//     root = insertRecursive(root, 5);
//     root = insertRecursive(root, 3);
//     root = insertRecursive(root, 8);
//     root = insertRecursive(root, 2);
//     root = insertRecursive(root, 4);
//     root = insertRecursive(root, 7);
//     root = insertRecursive(root, 9);

//     printf("Inorder Traversal:\n");
//     inorderRecursivePrint(root);

//     return 0;
// }
